src/lib/simulator/Simulator.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include "Particle.h"
+#include "ParticleContainer.h"
+#include "config/config.h"
+#include "simulator/io/Plotter.h"
+#include "utils/ArrayUtils.h"
+#include <cmath>
+#include <spdlog/sinks/basic_file_sink.h>
+#include <spdlog/sinks/stdout_color_sinks.h>
+#include <spdlog/spdlog.h>
+#include <utility>
+
+namespace simulator {
+/** <p> physics concept for force calculation </p>
+      * calculates the force for all particles,
+      *
+      * This is a concept since it allows the physics to be transparent for compiler optimizations.
+      * Trade off is binary size.
+      * \param T
+      * \param particle1
+      * \param particle2
+      */
+template<typename T>
+concept Physics = requires(T type, Particle const &particle1, Particle const &particle2) {
+  { T::calculate_force(particle1, particle2) } -> std::convertible_to<std::array<double, 3>>;
+};
+/**
+ * The main Simulator class. can be configured by providing a config and a plotter. Some methods use a physics model provided at compile time.
+ * */
+class Simulator {
+ private:
+  ParticleContainer particles;
+  std::unique_ptr<io::Plotter> plotter;
+  std::shared_ptr<config::Config> config;
+
+  double start_time;
+  double end_time;
+  double delta_t;
+
+ public:
+  /**
+   * \param particles the particle container
+   * \param plotter An instance plotter.
+   * \param config the runtime configuration
+   * */
+  explicit Simulator(
+      ParticleContainer particles,
+      std::unique_ptr<io::Plotter> plotter,
+      const std::shared_ptr<config::Config> &config);
+
+  /**
+  * @brief Runs the simulation.
+  *
+  * Executes the main simulation loop, updating particle positions, forces, and velocities
+  * until the end time is reached. Periodically plots the particles based on the IO interval.
+  */
+  template<Physics PY, bool IO>
+  auto run() -> void;
+
+  /*! <p> Function for position calculation </p>
+   *
+   * calculates the position for all particles, takes no arguments and has no return value
+   */
+  auto calculate_position() -> void;
+
+  /*! <p> Function for velocity calculation </p>
+  * calculates the velocity for all particles, takes no arguments and has no return value
+  */
+  auto calculate_velocity() -> void;
+
+  /**
+  * @brief Calculates forces between particles.
+  *
+  * Resets forces for all particles, then calculates and updates forces for each particle pair.
+  */
+  template<Physics PY>
```

Comment:
I think the concept of using physics as a template is interesting. I think it also makes sense here if you want to separate "gravitational" simulations from "molecular" simulation. 

However, this doesn't seem like an easy thing to extend in case you want to add a lot of different forces. More so, you may want to add different forces for same physical problem, like different potentials for MD simulations